source/slang/syntax.cpp

```diff
@@ -1070,7 +1068,7 @@ void Type::accept(IValVisitor* visitor, void* extra)
         return declRef.GetHashCode();
     }
 
-    Type* GenericDeclRefType::CreateCanonicalType()
+    RefPtr<Type> GenericDeclRefType::CreateCanonicalType()
```

Comment:
Okay, if I understand the ownership rules for canonical types now, I have to ask why these had to change to return `RefPtr<Type>`? If `this` is already canonical, we'll return `this` and nobody needs to increment the reference count (it must have been live if we called `GetCanonicalType()` on it), and if it wasn't canonical, we'd fill in its `canonicalTypeRefPtr` field, so that the canonical type is referenced, right?

I guess this isn't too important, since it is probably safest in our codebase to have functions return `RefPtr`s in general, to avoid reference-counting errors.

Reply:
If `this` is canonical, `Type::GetCanonicalType` will call `detach` function on the returned `RefPtr` to bring down that reference count back. This is me trying to bottleneck all potential calls to `detach` in  a single place. Otherwise, there could be multiple places in various `CreateCanonicalType` implementations that gets another `RefPtr<Type>` by calling some other functions and will need to detach them there...